### PR TITLE
Risk Model PR review

### PR DESCRIPTION
**Pull request by ZiqiaoLin**
Hi Charlie,

The PR is complete now. Please take a look and feel free to leave me any comments. 
I am cleaning up the example code and run the model with more factors and see the results

### CODE REVIEW

**Comment by charliereese** [comment]
Please move factor_utils.py to risk_model/utils/risk_utils.py

Please move factor_covariance_adjustment.py to risk_model/utils/risk_adjustments.py

**Comment by ZiqiaoLin** [comment]
> PR should only contain 3 files, not 13

I think it counts several files that you made changes yesterday. I will pull the newest version and then make the PR

File: investos/portfolio/risk_model/factor_risk.py [hunk 1]
```diff
@@ -1,23 +1,184 @@
...
+            self._pull_fos_risk_factor_dfs()
+            print("\n2. Cross Sectional Regression")
+            self._generate_risk_models()
+            print("\n3. Apply Risk Adjustment")
+            self._apply_risk_model_adjustments()
+        else:
+            self._factor_covariance = factor_covariance
+            self._factor_loadings = factor_loadings
+            self._idiosyncratic_variance = idiosyncratic_variance
+        #self._drop_excluded_assets()
```
Comment (charliereese) [inline investos/portfolio/risk_model/factor_risk.py:52]: No commented out code anywhere as discussed

File: investos/portfolio/risk_model/factor_risk.py [hunk 2]
```diff
@@ -1,23 +1,184 @@
...
+        self._fos_return_uuid = fos_return_uuid
+
+        self._end_date = end_date
+        self._start_date = start_date
+        self._adjustments = adjustments
+        #self._drop_excluded_assets()
+
+        if self._fos_risk_factor_uuids: 
+            print("=================Barra Risk Model================= ")
+            print("\n1. Initialization")
```
Comment (charliereese) [inline investos/portfolio/risk_model/factor_risk.py:42]: Print statements should be more descriptive of what is going on.

File: investos/portfolio/risk_model/factor_risk.py [hunk 3]
```diff
@@ -1,23 +1,184 @@
...
+        else:
+            self._factor_covariance = factor_covariance
+            self._factor_loadings = factor_loadings
+            self._idiosyncratic_variance = idiosyncratic_variance
+        #self._drop_excluded_assets()
+
+    # Initialization
+    def _pull_fos_risk_factor_dfs(self):
+        # pull from feature hub
+        fos.api_key = os.environ.get("FORECASTOS_API_KEY")                  #.env including my api_key
```
Comment (charliereese) [inline investos/portfolio/risk_model/factor_risk.py:57]: Should already be set - shouldn't be needed.

File: investos/portfolio/risk_model/factor_risk.py [hunk 4]
```diff
@@ -1,23 +1,184 @@
...
+        if not fos.api_key:
+            raise ValueError("Feature hub connection failed. Please check the API key is valid.")
+        dataframes = [fos.Feature.get(factor_uuids[factor]).get_df().rename(columns={'value': factor})
```
Comment (charliereese) [inline investos/portfolio/risk_model/factor_risk.py:60]: Poor logic for whether df has datetime column or not

File: investos/portfolio/risk_model/factor_risk.py [hunk 5]
```diff
@@ -1,23 +1,184 @@
...
+                       for factor in (self._fos_risk_factor_uuids + self._fos_return_uuid)  if factor != 'rbics']
+
+        if dataframes:
```
Comment (charliereese) [inline investos/portfolio/risk_model/factor_risk.py:63]: Shouldn't be needed.

File: investos/portfolio/risk_model/factor_risk.py [hunk 6]
```diff
@@ -1,23 +1,184 @@
...
+            df_merged = reduce(lambda left, right: pd.merge(left, right, how='left', on=['datetime', 'id']), dataframes)
+        else:
+            raise ValueError("factor loading is empty. Please check the risk factor uuids")
+        # special handling industry factors
+        if 'rbics' in self._fos_risk_factor_uuids:
```
Comment (charliereese) [inline investos/portfolio/risk_model/factor_risk.py:68]: Shouldn't reference specific risk factor, should be checking for datetime

File: investos/portfolio/risk_model/factor_risk.py [hunk 7]
```diff
@@ -1,23 +1,184 @@
...
+            df_industry = pd.get_dummies(fos.Feature.get(factor_uuids['rbics']).get_df().rename(columns={'value': 'industry'}),columns = ['industry'])
+            df_industry.iloc[:,1:] = df_industry.iloc[:,1:].astype(int)     # convert bool to int for industry exposure
+            df_merged = df_merged.merge(df_industry,how='left', on='id')
 
-        self.factor_covariance = factor_covariance
-        self.factor_loadings = factor_loadings
-        self.idiosyncratic_variance = idiosyncratic_variance
+            self._fos_risk_factor_uuids.remove('rbics')
```
Comment (charliereese) [inline investos/portfolio/risk_model/factor_risk.py:73]: Can't reference specific factor

File: investos/portfolio/risk_model/factor_risk.py (hunk 6 above)
Comment (ZiqiaoLin) [reply investos/portfolio/risk_model/factor_risk.py:68]: got it! Checking for col 'datetime' to decide how to merge

**Comment by charliereese** [comment]
Missing merge_asof logic as discussed

**Comment by charliereese** [comment]
> Missing merge_asof logic as discussed

(For filling factor data into all timesteps where data may need to be filled forward)

**Comment by charliereese** [comment]
Remove comments that shouldn't be there (like "fix numerical error")

**Comment by charliereese** [comment]
Need more descriptive variable names everywhere so people understand what is going on at each step

**Comment by charliereese** [comment]
Never start non-constant variable names with capital letter.

**Comment by charliereese** [comment]
Please close, process comments, and reopen. Please follow architecture we discussed.

**Comment by charliereese** [comment]
Please set up and run (ruff) pre-commit before next PR (and any future commits).

File: investos/portfolio/risk_model/factor_risk.py [hunk 8]
```diff
@@ -26,6 +187,8 @@ def _estimated_cost_for_optimization(self, t, w_plus, z, value):
 
         Not used to calculate simulated costs for backtest performance.
         """
+
+        # TODO: need to update this
```
Comment (charliereese) [inline investos/portfolio/risk_model/factor_risk.py:191]: Remove this comment

File: investos/portfolio/risk_model/factor_risk.py [hunk 9]
```diff
@@ -1,23 +1,184 @@
...
+        Generate idio returns
+        """
+        # Merge the DataFrames on 'date'
+        self._df_idio = pd.merge(self._df_loadings, self._df_factor_returns.reset_index(), on="datetime", suffixes=("", "_factor_returns"))
+
+        # Multiplying matching columns
+        for col in (self._fos_risk_factor_uuids + self._fos_industry_factors):
+            self._df_idio[f"calc_f_r_{col}"] = self._df_idio[col] * self._df_idio[f"returns_{col}"]
+
+        # # Dropping the extra columns
```
Comment (charliereese) [inline investos/portfolio/risk_model/factor_risk.py:144]: Double hashed comment

File: investos/portfolio/risk_model/factor_risk.py [hunk 10]
```diff
@@ -1,23 +1,184 @@
...
-        self.factor_loadings = factor_loadings
-        self.idiosyncratic_variance = idiosyncratic_variance
+            self._fos_risk_factor_uuids.remove('rbics')
+            self._fos_industry_factors = [col for col in df_industry.columns if col.startswith('industry_')]
+        
+        self._df_loadings =df_merged.replace([np.inf, -np.inf], np.nan).dropna()
+        self._df_loadings = self._df_loadings[(self._df_loadings.datetime >= self._end_date) & (self._df_loadings.datetime <= self._start_date)]
+        return
+    
+    def _generate_risk_models(self, timestamp = 'datetime') -> None:
```
Comment (charliereese) [inline investos/portfolio/risk_model/factor_risk.py:80]: Any function that is ~20 or greater lines should be split into new functions. Even +10 is a bit long.

**Comment by charliereese** [comment]
Don't use confusing abbreviations like FCM for factor covariance matrix (also reiterating to not using capitals for variable names).

File: investos/portfolio/risk_model/factor_risk.py (hunk 1 above)
Comment (ZiqiaoLin) [reply investos/portfolio/risk_model/factor_risk.py:52]: resolved

File: investos/portfolio/risk_model/factor_risk.py (hunk 2 above)
Comment (ZiqiaoLin) [reply investos/portfolio/risk_model/factor_risk.py:42]: resolved

File: investos/portfolio/risk_model/factor_risk.py (hunk 3 above)
Comment (ZiqiaoLin) [reply investos/portfolio/risk_model/factor_risk.py:57]: resolved

File: investos/portfolio/risk_model/factor_risk.py (hunk 4 above)
Comment (ZiqiaoLin) [reply investos/portfolio/risk_model/factor_risk.py:60]: resolved

File: investos/portfolio/risk_model/factor_risk.py (hunk 5 above)
Comment (ZiqiaoLin) [reply investos/portfolio/risk_model/factor_risk.py:63]: resolved

File: investos/portfolio/risk_model/factor_risk.py (hunk 6 above)
Comment (ZiqiaoLin) [reply investos/portfolio/risk_model/factor_risk.py:68]: resolved

File: investos/portfolio/risk_model/factor_risk.py (hunk 7 above)
Comment (ZiqiaoLin) [reply investos/portfolio/risk_model/factor_risk.py:73]: resolved

**Pull request closed without merging**